### PR TITLE
AlignDoubleArrowFixer - fix handling of nested array

### PR DESCRIPTION
**Pull request by SpacePossum**
Fixes: https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/1377

Replaces
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1417

on 1.10

### CODE REVIEW

File: Symfony/CS/Fixer/Contrib/AlignDoubleArrowFixer.php [hunk 1]
```diff
@@ -70,17 +71,10 @@ public function getDescription()
...
-        if (empty($startAt)) {
-            $startAt = 0;
-        }
-
-        if (empty($endAt)) {
-            $endAt = count($tokens);
-        }
-
         for ($index = $startAt; $index < $endAt; ++$index) {
+            /** @var Token $token */
```
Comment (keradus) [inline Symfony/CS/Fixer/Contrib/AlignDoubleArrowFixer.php]: is it really needed?
https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/1.11/Symfony/CS/Tokenizer/Tokens.php#L23

File: Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php [hunk 2]
```diff
@@ -30,6 +30,77 @@ public function provideFixCases()
     {
         return array(
             array(
+                '$array = array(
```
Comment (keradus) [inline Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php]:
```php
array(
    'asd',
),

File: Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php [hunk 3]
```diff
@@ -30,6 +30,77 @@ public function provideFixCases()
...
+                "closure" => function ($param1, $param2) {
+                    return;
+                }
+            );', ),
+            array(
+                "<?php
```
Comment (keradus) [inline Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php]: should be '

File: Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php [hunk 4]
```diff
@@ -30,6 +30,77 @@ public function provideFixCases()
...
+        ]);",
+                "<?php
+        return new JsonResponse([
+            'result' => 'OK',
+            'html' =>    renderView('views/my_view.html.twig', array(
+                'foo' => 'bar',
+                'foofoo' => 42,
+            )),
+        ]);",
+
```
Comment (keradus) [inline Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php]: not needed

File: Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php [hunk 5]
```diff
@@ -30,6 +30,77 @@ public function provideFixCases()
...
+                ),
+            array(
+                "<?php
+        return new JsonResponse([
+            'result' => 'OK',
+            'html'   => renderView('views/my_view.html.twig', [
+                'foo'    => 'bar',
+                'foofoo' => 42,
+            ]),
+            'baz'    => 'OK',
```
Comment (keradus) [inline Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php]: please remove alignment after baz for input and let see if output will be received

File: Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php [hunk 6]
```diff
@@ -30,6 +30,77 @@ public function provideFixCases()
...
+        ]);",
+
+                "<?php
+        return new JsonResponse([
+            'result' => 'OK',
+            'html' =>    renderView('views/my_view.html.twig', [
+                'foo' =>   'bar',
+                'foofoo' =>    42,
+            ]),
+            'baz'    => 'OK',
```
Comment (keradus) [inline Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php]: please remove alignment after baz for input and let see if output will be received

File: Symfony/CS/Fixer/Contrib/AlignDoubleArrowFixer.php (hunk 1 above)
Comment (SpacePossum) [inline Symfony/CS/Fixer/Contrib/AlignDoubleArrowFixer.php]: Sadly that line is missing on 1.10.
I'll remove this hint at the end when this PR is stable

File: Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php (hunk 4 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php]: I think it is, since it is the only test for long array with a short array in it starting on the same line.

Comment (keradus) [inline Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php]: the empty 79 line is not needed

Comment (SpacePossum) [inline Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php]: oops :)
will remove

File: Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php (hunk 2 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php]: I'm sorry, I don't understand what I should do here.

Comment (keradus) [inline Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php]: - missing PHP opening tag
- missing enter after comma to be consistent with other cases
- missing indent for array content

```php
            array(
                '<?php
    $array = array(
        "closure" => function ($param1, $param2) {
            return;
        }
    );',
            ),
```

File: Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php [hunk 7]
```diff
@@ -30,6 +30,74 @@ public function provideFixCases()
...
+            ),
+            array(
+                '<?php
+        return new JsonResponse([
+            "result" => "OK",
+            "html"   => renderView("views/my_view.html.twig", [
+                "foo"    => "bar",
+                "foofoo" => 42,
+            ]),
+            "baz" => "OK",
```
Comment (keradus) [inline Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php]: `OK` is not aligned :(

Comment (SpacePossum) [inline Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php]: hmm this PR is now in WIP state, it seems the is something off with detecting and/or settting the groups in the `AbstractAlignFixer`   : (

Comment (SpacePossum) [inline Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php]: Note to myself:
```php
<?php
    $data = [
        "foo"   ALIGNABLE1 => "Bar",
        "main"  ALIGNABLE1 => [array("baz"  ALIGNABLE3 => "Test")],
        "bar"   ALIGNABLE1 => array(),
    ];
    $data = array(
        "foo"   ALIGNABLE2 => "Bar",
        "Main"  ALIGNABLE2 => array("baz"  ALIGNABLE3 => "Test"),
        "bar"   ALIGNABLE2 => array(),
    );
    $var = [];
    foreach ($foo as $i => $bar) {
        $var[] = /* Comment */ [$i  ALIGNABLE4 => $bar];
    }
```

this is wrong on 1.10 and on the branch of this PR, the repeat of ALIGNABLE3 in two diff. groups causes the alignment to be done wrongly.

Comment (keradus) [inline Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php]: Or not
> align_double_arrow [contrib]
> Align double arrow symbols in ***consecutive lines***.

Comment (SpacePossum) [inline Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php]: hmm, maybe I misunderstand `consecutive` here.
For example:
```
<?php
    $data1 = [
        "foo"   => "Bar",
        "main1" => [array("baz" => "Test")],
        "bar"   => array(),
    ];

   // couple of 100s lines of code

    $data2 = [
        "foo"   => "Bar",
        "main2" => [array("bazaaaaaaaaaaaaaaaaaaaaaaaaa" => "Test")],
        "bar"   => array(),
    ];
```

do you really want to align data 1[ main1 ] with data 2[ main2 ] ?

Comment (keradus) [inline Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php]: That is what I said that not-aligning `baz` may not be an error ;)

Comment (SpacePossum) [inline Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php]: >That is what I said that not-aligning baz may not be an error ;)

but you were sad about it

>OK is not aligned :(

Comment (SpacePossum) [inline Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php]: anyway, I've cleaned up this PR as much as possible. Please let me know if I'm still missing cases here :)

Comment (keradus) [inline Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php]: I was sad, but still open-minded enought to see that the fixer not do it by design.
Thanks for update!

File: Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php [hunk 8]
```diff
@@ -31,11 +31,81 @@ public function provideFixCases()
...
+        ]),
+        "baz" => "OK",
+    ]);',
+            ),
+            array(
+                '<?php
+    switch ($a) {
+        case "prod":
+            break;
+    }
```
Comment (keradus) [inline Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php]: how is this case related to fixer?

File: Symfony/CS/Fixer/Contrib/AlignDoubleArrowFixer.php [hunk 9]
```diff
@@ -148,10 +140,12 @@ private function injectAlignmentPlaceholders(Tokens $tokens, $startAt = null, $e
             }
 
             if ($token->equals(',')) {
-                do {
+                for ($i = $index; $i < $endAt - 1; ++$i) {
+                    if ($tokens[$i + 1]->equals('[') || $tokens[$i + 1]->isGivenKind(T_ARRAY) || false !== strpos($tokens[$i - 1]->getContent(), "\n")) {
```
Comment (keradus) [inline Symfony/CS/Fixer/Contrib/AlignDoubleArrowFixer.php]: `$tokens->isArray($i + 1)`

**Comment by GrahamCampbell** [comment]
appveyor is broken here too

File: Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php (hunk 8 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php]: I dunno, I didn't add it https://github.com/SpacePossum/PHP-CS-Fixer/blob/1.10/Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php#L34, you want me to take it out?

Comment (keradus) [inline Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php]: ok, red section is on top, ignore me ;)

File: Symfony/CS/Fixer/Contrib/AlignDoubleArrowFixer.php (hunk 9 above)
Comment (SpacePossum) [inline Symfony/CS/Fixer/Contrib/AlignDoubleArrowFixer.php]: :+1:   done in 0e43781dcb37ba5421d1b744872f65c2f9c6da64

File: Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php (hunk 8 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/Fixer/Contrib/AlignDoubleArrowFixerTest.php]: I see, I'll move it back to the top to make the diff. smaller :)

**Comment by keradus** [comment]
Thank you @SpacePossum.

**Comment by SpacePossum** [comment]
\o/
thanks @keradus  :shipit:  :)

File: Symfony/CS/Fixer/Contrib/AlignDoubleArrowFixer.php [hunk 10]
```diff
@@ -148,10 +140,12 @@ private function injectAlignmentPlaceholders(Tokens $tokens, $startAt = null, $e
             }
 
             if ($token->equals(',')) {
-                do {
+                for ($i = $index; $i < $endAt - 1; ++$i) {
```
Comment (Soullivaneuh) [inline Symfony/CS/Fixer/Contrib/AlignDoubleArrowFixer.php]: @SpacePossum This `for` statement is the reason of https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/1450 issue.

Revert this part by the old while solve it. Any idea?

Comment (Soullivaneuh) [inline Symfony/CS/Fixer/Contrib/AlignDoubleArrowFixer.php]: More precisely, `$tokens->isArray($i + 1) || ` is the cause. Removing it solves the issue.

@SpacePossum This bug is very annoying. Could you fix it quickly? If not, I suggest a revert of this part and works for your related issue later.

cc @keradus
